doc/VERSION.md

```diff
@@ -2,47 +2,84 @@
 
 The version of Elektra is handled with the kdb.h macros
 `KDB_VERSION` which is a string and `KDB_VERSION_MAJOR`,
-`KDB_VERSION_MINOR` and `KDB_VERSION_MICRO` which are
-numbers. They represent the public announced version
-information.
+`KDB_VERSION_MINOR` and `KDB_VERSION_PATCH` which are
+numbers for the publicly announced versions.
 
-The same information can be retrieved at run-time using
+The version can also be retrieved at run-time from KDB:
 
 ```
 system/elektra/version/constants/KDB_VERSION
 system/elektra/version/constants/KDB_VERSION_MAJOR
-system/elektra/version/constants/KDB_VERSION_MICRO
 system/elektra/version/constants/KDB_VERSION_MINOR
+system/elektra/version/constants/KDB_VERSION_PATCH
 ```
 
-This is the API to programs using Elektra. Its interface
-is defined in [src/include/kdb.h](/src/include/kdb.h.in).
-Both applications and plugins use this API.
-
-Additionally there is also a very small API
-to plugins. It consists of only 5 functions
-and is described in [src/plugins/doc/doc.c](/src/plugins/doc/doc.c).
+## Scope
+
+The version applies to following parts of Elektra:
+
+- the API for programs using Elektra. Its interface
+  is declared in [src/include/kdb.h](/src/include/kdb.h.in).
+  Both applications and plugins use this API.
+- the high-level API as declared in
+  [src/include/elektra.h](/src/include/elektra.h).
+- the API to plugins as declared in
+  [src/plugins/doc/doc.h](/src/plugins/doc/doc.h).
+- the API to [libease](/src/include/kdbease.h)
+- the API to [libmeta](/src/include/kdbmeta.h)
+- the API to [libmerge](/src/include/kdbmerge.h)
+- the API to [module loading](/src/include/kdbmodule.h)
+- the API to [libnotification](/src/include/kdbnotification.h)
+- the API to [libopts](/src/include/kdbopts.h)
+- plugins that have `infos/status` of `compatible`.
+- the CLI tool `kdb` with its command-line
+  arguments, KDB access and its return values.
+
+## Behavior
+
+The following order defines the behavior of Elektra:
+
+1. [ABI](/tests/abi) test cases
+2. Non-ABI test cases (including Shell Recorder)
+3. The [API documentation](https://doc.libelektra.org/api/latest/html/)
+4. The [man pages](/doc/help)
+5. Tutorials (excluding Shell Recorder)
+6. Examples
+7. Implementation
```

Comment:
A few more comments on this list:

As mentioned below, IMO the implementation should not be part of the list. AFAIK this list defines the *intended behaviour* of Elektra. Putting the implementation on this list makes all bugs part of the intended behaviour. The implementation is Elektra's *actual behaviour*, which should of course match the *intended behaviour*, but not be part of its definition.

In other words, any behaviour that is undocumented (docu or examples) and untested (tests), outside of its implementation should be considered a private implementation detail for which we make no guarantees.

The points regarding test cases should be restricted to the parts of the tests that test public API function. For example there are test cases, which test the internal key unescape and key escape functions. Since these are not part of the public API, we should not make any guarantees for their behaviour (as long as public functions remain unaffected).

Reply:
As a customer of Elektra (a developer of an elektrified program) I would be most interested in the API docu. I would usually not read through the ABI test cases to understand the intention of the API.

Internally this list makes sense, as the implementation obviously defines the actual behavior. Maybe I just misinterpreted the intention of the list.

Reply:
> The implementation is Elektra's actual behaviour, which should of course match the intended behaviour, but not be part of its definition.

You are completely right. I fixed it in d13860247e5dd85aba9cdc107b228987fc5cf609. I had a different starting sentence before (about where to fix the bugs, and then implementation is on the list).

> The points regarding test cases should be restricted to the parts of the tests that test public API function. For example there are test cases, which test the internal key unescape and key escape functions. Since these are not part of the public API, we should not make any guarantees for their behaviour (as long as public functions remain unaffected).

Difficult to say without any actual bug at hand. But what is clear that these test cases are subject to be changed if the documentation says otherwise. If the documentation said nothing before, we might decide to fix the documentation. And then, according to this proposed list, the docu would win against these test cases anyway.

Reply:
(GitHub did not show the comment of @mpranj thats why I answer now)

> As a customer of Elektra (a developer of an elektrified program) I would be most interested in the API docu. I would usually not read through the ABI test cases to understand the intention of the API.

Yes, I agree.

> Internally this list makes sense, as the implementation obviously defines the actual behavior. Maybe I just misinterpreted the intention of the list.

No, you interpreted correctly and the order in the list is up to discussion. We can swap these points. My rationale was that it is very easy to verify if the implementation behaves according the test cases and very difficult (impossible) to verify if it behaves according to sentences in natural languages. Thus I think it is more practical to adjust the sentences and not the test cases.

Of course, this is not so nice for application developers that implemented applications according to wrong documentation. We can improve the situation by including more test cases in the documentation.

Another reason for as-is is that applications with special requirements can, if the developers are serious about it, contribute test cases. Then they have the guarantee that we will only adapt docu and not their test cases (which would make their application fail).